pkg/os/volume/api.go

```diff
@@ -268,6 +272,96 @@ func getTarget(mount string) (string, error) {
 	return volumeString, nil
 }
 
+// GetVolumeIDFromTargetPath returns the volume id of a given target path.
+func (VolumeAPI) GetClosestVolumeIDFromTargetPath(targetPath string) (string, error) {
+	volumeString, err := findClosestVolume(targetPath)
+
+	if err != nil {
+		return "", fmt.Errorf("error getting the closest volume for the path=%s, err=%v", targetPath, err)
+	}
+
+	return volumeString, nil
+}
+
+// findClosestVolume finds the closest volume id for a given target path
+// by following symlinks and moving up in the filesystem, if after moving up in the filesystem
+// we get to a DriveLetter then the volume corresponding to this drive letter is returned instead.
+func findClosestVolume(path string) (string, error) {
+	candidatePath := path
+
+	// run in a bounded loop to avoid doing an infinite loop
+	// while trying to follow symlinks
+	for i := 0; i < 256; i += 1 {
+		fi, err := os.Lstat(candidatePath)
+		if err != nil {
+			return "", err
+		}
+		isSymlink := fi.Mode()&os.ModeSymlink != 0
+
+		if isSymlink {
+			target, err := dereferenceSymlink(candidatePath)
+			if err != nil {
+				return "", err
+			}
+			// if it has the form Volume{volumeid}\ then it's a volume
+			if strings.HasPrefix(target, "Volume") {
```

Comment:
I feel the Volume prefix check can be tightened with a regex to exactly match `Volume{...}` format. If a disk layout has symlink chains with links named `VolumeData1` and `VolumeData2` that are actually links to `Volume{A}` and `Volume{B}`, the `HasPrefix` logic would return incorrect results.